src/openMVG/sfm/pipelines/sequential/sequential_SfM.cpp

```diff
@@ -1076,124 +1076,121 @@ bool SequentialSfMReconstructionEngine::Resection(const size_t viewIndex)
-              pose_J.depth(X_euclidean) > 0 &&
-              residual_I.norm() < std::max(4.0, map_ACThreshold_.at(I)) &&
-              residual_J.norm() < std::max(4.0, map_ACThreshold_.at(J)))
-            {
-              {
-                // Add a new track
-                Landmark & landmark = sfm_data_.structure[trackId];
-                landmark.X = X_euclidean;
-                landmark.obs[I] = Observation(xI, track.at(I));
-                landmark.obs[J] = Observation(xJ, track.at(J));
-                ++new_added_track;
-              } // critical
-            } // 3D point is valid
-          } // else (New 3D point)
-        }
-      }// For all correspondences
-/*
-#ifdef OPENMVG_USE_OPENMP
-      #pragma omp critical
-#endif
-      if (!map_tracksCommonIJ.empty())
-      {
-        std::cout
-          << "\n--Triangulated 3D points [" << I << "-" << J << "]:"
-          << "\n\t#Track extented: " << extented_track
-          << "\n\t#Validated/#Possible: " << new_added_track << "/" << new_putative_track
-          << "\n\t#3DPoint for the entire scene: " << sfm_data_.GetLandmarks().size() << std::endl;
-      }
-*/
-    }
+    // Go through each track in the new view and either add observations or new 3D points
+    for (const std::pair< size_t, tracks::submapTrack >& trackIt : map_tracksCommon)
+	{
+      const size_t trackId = trackIt.first;
+      const tracks::submapTrack & track = trackIt.second;
+	  // Position of the point in view I
+      const Vec2 xI = features_provider_->feats_per_view.at(I)[track.at(I)].coords().cast<double>();
+
+		// If the track was already reconstructed
+		if (sfm_data_.structure.count(trackId) != 0)
+		{
+		  // 3D point triangulated before, only add image observation if needed
+		  {
+			Landmark & landmark = sfm_data_.structure[trackId];
+			if (landmark.obs.count(I) == 0)
+			{
+			  const Vec2 residual = cam_I->residual(pose_I, landmark.X, xI);
+			  if (pose_I.depth(landmark.X) > 0 && residual.norm() < std::max(4.0, map_ACThreshold_.at(I)))
+			  {
+				landmark.obs[I] = Observation(xI, track.at(I));
+			  }
+			}
+		  }
+		}
+		else
+		{
+			// Go through the views that observe this track
+			const tracks::submapTrack & allViews_of_track = map_tracks_[trackId];
+			std::set<IndexT> new_track_observations_valid_views;
+
+			for (const std::pair< IndexT, IndexT >& trackViewIt : allViews_of_track)
+			{
+				const IndexT & J = trackViewIt.first;
+				// If view is valid try triangulation
+				if(J!=I && valid_views.count(J) != 0 )
+				{
+					const View * view_J = sfm_data_.GetViews().at(J).get();
+					const IntrinsicBase * cam_J = sfm_data_.GetIntrinsics().at(view_J->id_intrinsic).get();
+					const Pose3 pose_J = sfm_data_.GetPoseOrDie(view_J);
+					const Vec2 xJ = features_provider_->feats_per_view.at(J)[allViews_of_track.at(J)].coords().cast<double>();
+
+					// If successfuly triangulated add the observation from J view
+					if (sfm_data_.structure.count(trackId) != 0){
```

Comment:
I think this IF part can be simplified to
```
new_track_observations_valid_views.insert(J);
```
And so the point observation will be added by your loop below.
It will make the code simpler.